### PR TITLE
✨ ORC Image API update

### PR DESCRIPTION
**Pull request by mdbooth**
Updates the ORC Image API in light of a review conducted with @pierreprinetti in preparation for adding more ORC controllers. Primarily:

* Mechanical move of OpenStack resource-specific spec and status fields into `spec.resource` and `status.resource` respectively.
* Removes controllerOptions.
This functionality is replaced by `managementPolicy` and `managedOptions`.
* Add `import` for explicit import of an existing OpenStack resource.
Resource import was previously implied by resource spec fields. It is now explicit in its own field.
* Use field indexer for dependency trigger
No longer use labels for this. This reduces the API surface, and also removes an API restriction on allowed name length.
* Rename sourceURL to Download and remove sourceType
This mechanical change better reflects the purpose of this field, and simplifies the type slightly given that we may never add support for a `copy` source. It remains possible to add a `copy` source if required.

### CODE REVIEW

**Comment by EmilienM** [comment]
/assign pierreprinetti 

@pierreprinetti, I think @mdbooth wants your /lgtm before merging it.

**Review by pierreprinetti — COMMENTED** [review]
Amazing work, thank you @mdbooth for this.

I have one further (very light) concern about the mention of Glance in the godoc, in error messages and in exported identifiers. In my understanding of OpenStack, Glance is an implementation of the OpenStack Image API. While we surely want to comply with the actual thing in code, don't we want to externally refer to e.g. "OpenStack image" rather than "Glance image"?

If not, please consider running a quick `:%s/glance/Glance/gc`.

File: orc/api/v1alpha1/controller_options.go [hunk 1]
```diff
@@ -16,75 +16,50 @@ limitations under the License.
...
-// the default, AdoptOrCreate, if it is not set.
-func (o *ControllerOptions) GetOnCreate() ControllerOptionsOnCreate {
-	const defaultOnCreate = ControllerOptionsOnCreateAdoptOrCreate
-
-	if o == nil || o.OnCreate == nil {
-		return defaultOnCreate
-	}
-
-	return *o.OnCreate
+	DeletePolicy DeletePolicy `json:"deletePolicy,omitempty"`
```
[the remark this answers is not in the document]
Comment (mdbooth) [reply orc/api/v1alpha1/controller_options.go:55]: You can thank the fine developers of ASO for `detach`: I liked their word :) (Although I like that we haven't put it in an annotation better.)

Hmm... yes `onDelete` is probably better and it means basically the same thing. I primarily didn't want a boolean here, leaving us the option to add things we haven't thought of yet. I'll update this.

File: orc/api/v1alpha1/credentials_ref.go [hunk 2]
```diff
@@ -18,14 +18,18 @@ package v1alpha1
...
 type CloudCredentialsReference struct {
-	// Name is the name of a secret in the same namespace as the resource being provisioned.
+	// SecretName is the name of a secret in the same namespace as the resource being provisioned.
 	// The secret must contain a key named `clouds.yaml` which contains an OpenStack clouds.yaml file.
 	// The secret may optionally contain a key named `cacert` containing a PEM-encoded CA certificate.
 	// +kubebuilder:validation:Required
-	Name string `json:"name"`
+	// +kubebuilder:validation:MinLength:=1
+	// +kubebuilder:validation:MaxLength:=253
+	SecretName string `json:"secretName"`
```
[the remark this answers is not in the document]
Comment (mdbooth) [reply orc/api/v1alpha1/credentials_ref.go:27]: It's more explicit. The compatibility is in the 1-1 mapping of fields. What they're called isn't actually important for compatibility.

File: orc/api/v1alpha1/image_types.go [hunk 3]
```diff
@@ -323,73 +302,132 @@ type ImageSpec struct {
...
-var _ ObjectWithControllerOptions = &Image{}
+// ImageResourceStatus represents the observed state of a glance image
+type ImageResourceStatus struct {
+	// Status is the image status as reported by glance
+	// +optional
+	Status *string `json:"status,omitempty"`
 
-func (i *Image) GetControllerOptions() *ControllerOptions {
-	return i.Spec.ControllerOptions
+	// Hash is the hash of the image data published by glance
```
[the remark this answers is not in the document]
Comment (mdbooth) [reply orc/api/v1alpha1/image_types.go:386]: Yeah, sounds useful. I could get very ranty about this, though 💢 I've tried to keep it professional.

File: orc/api/v1alpha1/image_types.go [hunk 4]
```diff
@@ -323,73 +302,132 @@ type ImageSpec struct {
...
+// ImageImport specifies an existing image which will be imported instead of
+// creating a new image
+// +kubebuilder:validation:MinProperties:=1
+// +kubebuilder:validation:MaxProperties:=1
+type ImageImport struct {
+	// ID contains the UUID of an existing glance image. Note that when
+	// specifying an image import by ID, the image MUST already exist. The Image
+	// will enter an error state if the image does not exist.
+	// +optional
+	// +kubebuilder:validation:Format:=uuid
```
[the remark this answers is not in the document]
Comment (mdbooth) [reply orc/api/v1alpha1/image_types.go:330]: I'm sure I've found the canonical list before. I can't find it now, but I did find this GH issue documenting what appears to be the same list in a context which I think is the same, but if not then it's at least closely related: https://github.com/kubernetes/kubernetes/issues/97733

Note that `uuid` is exactly what you describe, with separate identifiers for versioned uuids.

**Comment by mdbooth** [comment]
I'm not convinced about 'OpenStack image', specifically when referring to configuration that's most definitely specific to glance. I'm not going to make that change now, but I'm receptive to being reeducated in the future.

**Comment by pierreprinetti** [comment]
There still are non-capitalised references to Glance in godocs and error messages. Besides that, LGTM :+1: 

I'll gladly apply the label once you rebase and possibly fix the capitalisation.

**Comment by mdbooth** [comment]
> There still are non-capitalised references to Glance in godocs and error messages. Besides that, LGTM 👍

Haha, I intentionally lowercased them 😆 I can put it back.

Interested to know where the merge conflict came from...

**Comment by pierreprinetti** [comment]
/lgtm

:rocket:

**Comment by pierreprinetti** [comment]
There is a bunch of other occurrences of uncapitalised Glance (e.g. in internal/controllers/image/reconcile.go) by the way.

File: orc/api/v1alpha1/credentials_ref.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pierreprinetti) [reply orc/api/v1alpha1/credentials_ref.go:27]: I get that the important bit is to be able to reuse the same secret and to have a similar auth mechanism. What I'm saying is that having different field names in different objects pointing to the same thing may be confusing.

**Pull request merged**